keps/sig-api-machinery/20190226-network-proxy.md

```diff
@@ -0,0 +1,519 @@
+listening at 1.2.3.5:5679. While these are represented as resources
+they are not intended to be loaded dynamically.
+The KAS loads them as configuration at start time.
+
+```yaml
+apiVersion: connectivityservice.k8s.io/v1alpha1
+kind: ConnectivityServiceConfiguration
+connectionService:
+  name: direct
+  connection:
+    type: direct
+---
+apiVersion: connectivityservice.k8s.io/v1alpha1
+kind: ConnectivityServiceConfiguration
+connectionService:
+  name: Master
+  connection:
+    type: grpc
+    url: grpc://1.2.3.4:5678
+    caBundle: file1.pem
+    clientKeyFile: proxy-client1.key
+    clientCertFile: proxy-client1.crt
+---
+apiVersion: connectivityservice.k8s.io/v1alpha1
+kind: ConnectivityServiceConfiguration
+connectionService:
+  name: Cluster
+  connection:
+    type: grpc
+    url: grpc://1.2.3.5:5679
+    caBundle: file2.pem
+    clientKeyFile: proxy-client2.key
+    clientCertFile: proxy-client2.crt
+```
+
+NetworkContext could be extended to contain more contextual information.
+This would allow smarter routing based on the k8s object KAS is processing
+or which user/tenant tries to initiate the request, etc.
+
+### Proxy gRPC definition
+
+In order to serve a proxy request, one gRPC bidirectional stream on proxy
+server is created to serve it. It's a 1:1 mapping from TCP connection to a
+gRPC stream, so the state of TCP connection is exactly the same as the gRPC
+stream state.
+
+```grpc
+syntax = "proto3";
+
+service ProxyService {
+  // Proxy a TCP connection to a remote address defined by ConnectParam.
+  // The ConnectParam is defined in metadata under key "x-kube-net-proxy".
+  // metadata["x-kube-net-proxy"] = base64.Encode(proto.Marshal(connectOptions))
+  rpc Proxy(stream Payload) returns (stream Payload) {}
+}
+
+// ConnectOptions defines the remote TCP endpoint to connect
+message ConnectOptions {
+  string remote_addr = 1; // remote address to connect to. e.g. 8.8.8.8:53
+}
+
+// Payload defines a TCP payload.
+message Payload {
+  bytes data = 1;
+}
+```
+
+### Connectivity Proxy
+
+The connectivity proxy(s) can run in the same container as the KAS.
+It should run on the same machine and must run in the same flat network as the KAS.
+It listens on a port for gRPC connections from the KAS.
+This port would be for forwarding traffic to the appropriate cluster.
+It should have an administrative port speaking https.
+The administrative port serves the liveness probe and metrics.
+The liveness probe prevents a partially broken cluster
+where the KAS cannot connect to the cluster. This port also serves
+pprof debug commands and monitoring data for the proxy.
+
+### Direct Connection
+
+This connection type uses the default dialer.
+This allows use of the connectivity service without the connectivity proxy.
+This is a quick way to run the system in a “legacy” or fallback mode.
+Simple clusters (not needing network segregation) run this way to avoid the overhead
+(in latency or configuration) of the connectivity proxy.
+
```

Comment:
There is a missing details on how a `cluster network` will initiate connection to `connectivity proxy`. How will it work with multiple cluster networks, multiple connectivity proxy, or both.

Reply:
So I've deliberately avoided dictating a solution for the connect back issue in this KEP. I agree that Google, Amazon and others will want to solve this problem. In fact I believe that there is (will be) a small sub-project right now being started up in sig-cloud-provider/sig-networking to create a reference implementation for solving this issue. Happy to provide more details on that as it become solidified. However I would like to keep the alpha release of this KEP as simple as possible.